banyand/liaison/grpc/discovery.go

```diff
@@ -37,48 +37,74 @@ import (
 var errNotExist = errors.New("the object doesn't exist")
 
 type discoveryService struct {
-	metadataRepo metadata.Repo
-	nodeRegistry NodeRegistry
-	groupRepo    *groupRepo
-	entityRepo   *entityRepo
-	log          *logger.Logger
-	kind         schema.Kind
+	metadataRepo    metadata.Repo
+	nodeRegistry    NodeRegistry
+	groupRepo       *groupRepo
+	entityRepo      *entityRepo
+	shardingKeyRepo *shardingKeyRepo
+	log             *logger.Logger
+	kind            schema.Kind
 }
 
 func newDiscoveryService(kind schema.Kind, metadataRepo metadata.Repo, nodeRegistry NodeRegistry) *discoveryService {
-	sr := &groupRepo{resourceOpts: make(map[string]*commonv1.ResourceOpts)}
+	gr := &groupRepo{resourceOpts: make(map[string]*commonv1.ResourceOpts)}
 	er := &entityRepo{entitiesMap: make(map[identity]partition.EntityLocator)}
+	sr := &shardingKeyRepo{shardingKeysMap: make(map[identity]partition.ShardingKeyLocator)}
 	return &discoveryService{
-		groupRepo:    sr,
-		entityRepo:   er,
-		kind:         kind,
-		metadataRepo: metadataRepo,
-		nodeRegistry: nodeRegistry,
+		groupRepo:       gr,
+		entityRepo:      er,
+		shardingKeyRepo: sr,
+		kind:            kind,
+		metadataRepo:    metadataRepo,
+		nodeRegistry:    nodeRegistry,
 	}
 }
 
 func (ds *discoveryService) initialize() error {
 	ds.metadataRepo.RegisterHandler("liaison", schema.KindGroup, ds.groupRepo)
 	ds.metadataRepo.RegisterHandler("liaison", ds.kind, ds.entityRepo)
+	if ds.kind == schema.KindMeasure {
+		ds.metadataRepo.RegisterHandler("liaison", ds.kind, ds.shardingKeyRepo)
+	}
 	return nil
 }
 
 func (ds *discoveryService) SetLogger(log *logger.Logger) {
 	ds.log = log
 	ds.groupRepo.log = log
 	ds.entityRepo.log = log
+	ds.shardingKeyRepo.log = log
 }
 
 func (ds *discoveryService) navigate(metadata *commonv1.Metadata, tagFamilies []*modelv1.TagFamilyForWrite) (pbv1.Entity, pbv1.EntityValues, common.ShardID, error) {
 	shardNum, existed := ds.groupRepo.shardNum(metadata.Group)
 	if !existed {
 		return nil, nil, common.ShardID(0), errors.Wrapf(errNotExist, "finding the shard num by: %v", metadata)
 	}
-	locator, existed := ds.entityRepo.getLocator(getID(metadata))
+	id := getID(metadata)
+	entityLocator, existed := ds.entityRepo.getLocator(id)
 	if !existed {
-		return nil, nil, common.ShardID(0), errors.Wrapf(errNotExist, "finding the locator by: %v", metadata)
+		return nil, nil, common.ShardID(0), errors.Wrapf(errNotExist, "finding the entity locator by: %v", metadata)
+	}
+	entity, entityValues, shardID, err := entityLocator.Locate(metadata.Name, tagFamilies, shardNum)
+	if err != nil {
+		return nil, nil, common.ShardID(0), err
+	}
+
+	if _, ok := ds.shardingKeyRepo.shardingKeysMap[id]; ok {
```

Comment:
Would you mind removing this line? `getLocator` can tell you whether the id exists.